File: node.py
from random import randint

class Node:
    #parents format: array of ids example: [1,2]
    #cpt format example: {['true', 'true']: 0.2
    #the first element in nature is true used in cpt
    def __init__(self,name, characters={}, parents = [], cpt={}, nature = []):
        self.name = name
        #dictionary for root node, is a string : float array
        self.characters = characters
        #list of node the parents
        self.parents = parents
        #dictionary of string : float
        self.cpt = cpt
        self.nature = nature
        self.status = ''
    
    def __hash__(self):
        return self.name
    
    def setStatus(self, status):
        self.status = status
    
    def translateParent(self):
        str = ''
        for p in self.parents:
            str+=p.status
        
        return str
    
    #this function should only be used to initialize the root node based on its characters table
    def startRoot(self):
        i = randint(0, 1000) / 1000
        prev = 0.0
        for key, value in self.characters.items():
            if i >= prev and i < value + prev:
                self.status = key
                return
            prev = value + prev
        

    

File: c_ver/node.cpp
#include "node.h"
#include <cstdlib>

Node::Node(std::string name, std::map<std::string, double> characters, std::vector<Node*> parents, std::map<std::string, double> cpt, std::pair<std::string, std::string> nature) {
    this->name = name;
    this->characters = characters;
    this->parents = parents;
    this->cpt = cpt;
    this->nature = nature;
}

std::string Node::getName() {
    return this->name;
}

std::string Node::getStatus() {
    return this->status;
}

std::string Node::translateParent() {
    std::string result;
    for (Node* node: this->parents) {
        result += node->getStatus();
    }
    return result;
}

void Node::setStatus(std::string status) {
    this->status = status;
}

void Node::startRoot() {
    double i = (double) rand() / RAND_MAX;
    double prev = 0;
    for (auto const& x: characters) {
        if ((i >= prev) && (i <= x.second)) {
            setStatus(x.first);
            return;
        }
        prev += x.second;
    }
}

void Node::startNode() {
    double i = (double) rand() / RAND_MAX;
    if (i <= cpt[translateParent()]) {
        setStatus(nature.first);
    }
    else {
        setStatus(nature.second);
    }
}

